go/fury/meta/meta_string_encoder.go

```diff
@@ -0,0 +1,261 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package meta
+
+import (
+	"fmt"
+	"unicode"
+)
+
+type Encoder struct {
+	specialChar1 byte
+	specialChar2 byte
+}
+
+func NewEncoder(specialCh1 byte, specialCh2 byte) *Encoder {
+	return &Encoder{
+		specialChar1: specialCh1,
+		specialChar2: specialCh2,
+	}
+}
+
+// Encode the input string to MetaString using adaptive encoding
+func (e *Encoder) Encode(input string) MetaString {
+	encoding := e.ComputeEncoding(input)
+	return e.EncodeWithEncoding(input, encoding)
+}
+
+// EncodeWithEncoding Encodes the input string to MetaString using specified encoding.
+func (e *Encoder) EncodeWithEncoding(input string, encoding Encoding) MetaString {
+	if len(input) > 32767 {
+		panic("Long meta string than 32767 is not allowed")
+	}
+	// execute encoding algorithm according to the encoding mode
+	length := len(input)
+	switch encoding {
+	case LOWER_SPECIAL:
+		return MetaString{
+			inputString:  input,
+			encoding:     encoding,
+			specialChar1: e.specialChar1,
+			specialChar2: e.specialChar2,
+			outputBytes:  e.EncodeLowerSpecial(input),
+			numChars:     length,
+			numBits:      length * 5,
+		}
+	case LOWER_UPPER_DIGIT_SPECIAL:
+		return MetaString{
+			inputString:  input,
+			encoding:     encoding,
+			specialChar1: e.specialChar1,
+			specialChar2: e.specialChar2,
+			outputBytes:  e.EncodeLowerUpperDigitSpecial(input),
+			numChars:     length,
```

Comment:
I removed the needs for numChars/numBits in #1565 , maybe we can simplify the implement a little.

Reply:
okay, I'll read and implement it.